### PR TITLE
refactor(typescript): packages/rspack/src/lib from js to ts and esm

### PR DESCRIPTION
**Pull request by shulaoda**
## Summary
Related to #7225 

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Comment by SoonIter** [comment]
@shulaoda could you check your qq email? 😄 @Timeless0911 sends a email to you

**Comment by shulaoda** [comment]
Ready to merge in? If ready, I will update the branch. @SoonIter

**Comment by SoonIter** [comment]
> Ready to merge in? If ready, I will update the branch. @SoonIter

reviewing, on the way

**Comment by SoonIter** [comment]
https://github.com/web-infra-dev/rspack/pull/7289/files/1bc8798593f2c93678fff2b7ea1e22dfaaeabbe5..2e4398453357dafad2a9557c64f77d40a1c61ebc#diff-5eba14dafadd5b7ee447411919692ad06f841b6b93770e131cc6b364771f68b5

😉OK, it looks good now

File: packages/rspack/src/lib/EnvironmentPlugin.ts [hunk 1]
```diff
@@ -1,50 +1,52 @@
...
-	// @ts-expect-error
-	constructor(...keys) {
+	keys: string[];
+	defaultValues: Record<string, string>;
+
+	constructor(...keys: string[] | [Record<string, string> | string]) {
 		if (keys.length === 1 && Array.isArray(keys[0])) {
 			this.keys = keys[0];
 			this.defaultValues = {};
-		} else if (keys.length === 1 && keys[0] && typeof keys[0] === "object") {
```
Comment (SoonIter) [inline packages/rspack/src/lib/EnvironmentPlugin.ts:23]: `[null].length === 1`, so... this modification might not be the same as before
just a corner case

File: packages/rspack/src/lib/LoaderTargetPlugin.ts [hunk 2]
```diff
@@ -1,41 +1,35 @@
...
-class LoaderTargetPlugin {
+export class LoaderTargetPlugin {
 	/**
-	 * @param {string} target the target
+	 * @param target the target
 	 */
-	constructor(target) {
-		this.target = target;
-	}
+	constructor(public readonly target: string) {}
```
Comment (SoonIter) [inline packages/rspack/src/lib/LoaderTargetPlugin.ts:18]: I think we can use `target: Target` rather than `as any` ?

**Review by SoonIter — APPROVED** [review]

**Comment by SoonIter** [comment]
thanks~ 🫰🏻

**Pull request merged**